### ISSUE TITLE
Validity of Box<T>

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/unsafe-code-guidelines#145]
What is the validity invariant of `Box<T>`? If this was just a library type it would be the same as `Unique<T>`, which is the same as `NonNull<T>`.  But actually most of the compiler treats `Box<T>` a lot like a reference, so at least de-facto right now, `Box<T>` likely has the same validity invariant as `&T` and `&mut T` (which is discussed in https://github.com/rust-lang/unsafe-code-guidelines/issues/76 and https://github.com/rust-lang/unsafe-code-guidelines/issues/77).

**Comment by gnzlbg** [comment]
I think that the validity invariant of `Box` is the same as `NonNull<T>`. 

The safety invariant of `Box` might be different.

**Comment by rkruppe** [comment]
`Box`es get annotated with attributes/metadata like `noalias`, `align`, `dereferenceable`, etc. which makes it immediate UB to e.g. pass an unaligned `Box` or a dangling `Box` around. Treating Box as just a non-null raw pointer is not enough to justify that.

**Comment by gnzlbg** [comment]
Thanks for pointing that out.

Summary from Discord chat with @rkruppe: Currently we generate `noalias` for `Box`, and `noalias` is only allowed on function arguments and return values in LLVM-IR. This means that if we want to generate the same LLVM-IR that we generate today, the validity invariant for `Box` needs to be pretty much that of `&mut T`. 

We also talked about whether it would be possible to enforce `noalias` on access to the Box only, (e.g. `Deref`/`DerefMut` could cast the `NonNull<T>` to a `&mut T`, and dereference that). This would allow relaxing the validity invariant of Box, and make `noalias` and the other properties part of its safety invariant. But this does not appear to be possible with LLVM today, and might never be. Relaxing the validity invariant does not enable users to do anything new, since they can always start using `NonNull`, and only upgrade to `Box` when the validity invariant is satisfied.

**Comment by RalfJung** [comment]
We really emit `dereferencable`? That's interesting, considering that the `Box` might well be deallocated during the functions runtime! For this reason, Stacked Borrows does *not* add a "protector" for `Box` (formerly "barrier", that's the mechanism that Stacked Borrows uses to justify the `dereferencable` attribute by making sure that the reference *remains* dereferencable during the entire runtime of the function).

**Comment by gnzlbg** [comment]
> We really emit dereferencable? 

@RalfJung This (https://rust.godbolt.org/z/E6Sl8a):

```rust
fn foo(x: Box<i32>) -> i32 { *x }
```

produces

```llvm
define i32 @foo(i32* noalias align 4 dereferenceable(4)) unnamed_addr #0 {
start:
  %1 = load i32, i32* %0, align 4
  %2 = bitcast i32* %0 to i8*
  tail call void @__rust_dealloc(i8* nonnull %2, i64 4, i64 4) #1
  ret i32 %1
}
```

**Comment by nico-abram** [comment]
The godbolt link doesn't seem to produce dereferenceable any more

**Comment by RalfJung** [comment]
Yes, that was changed with https://github.com/rust-lang/rust/pull/66645.

**Comment by nico-abram** [comment]
Just wanted to mention here that the validity invariants of Box<T> seem to contradict an assumption evmap had for shallow copying Box<T>: https://github.com/jonhoo/rust-evmap/issues/74

It even has a derive for it's ShallowCopy trait which could be used on aggregates with Box<T> fields

**Comment by RalfJung** [comment]
I think that issue is less about the validity invariant and more about the aliasing requirements.

**Comment by nico-abram** [comment]
You're completely right, I was a bit confused because `noalias` and `dereferenceable` were mentioned earlier in this thread

I was under the impression "language UB" was validity invariants (https://github.com/rust-lang/unsafe-code-guidelines/issues/77#issuecomment-706553237). Are aliasing (And I suppose providence) requirements considered separate from validity and safety invariants?

**Comment by RalfJung** [comment]
> Are aliasing  requirements considered separate from validity and safety invariants?

Yes, I consider them separate. I think they will be clearly separate in an eventual *specification* of Rust. However, I can see how for *teaching* purposes, that can be confusing...

> (And I suppose providence)

("provenance", not "providence" -- yeah these are all weird words.^^)

**Comment by jonhoo** [comment]
Yeah, I think that issue comes down to whether it is allowed to alias a `Box` as long as you a) do not violate `&mut`/`&` aliasing rules when dereferencing and b) only drop the `Box` once all the aliases are gone. Given that `Box` currently passes `noalias` even when you do not dereference, it sounds like that's not the case. Is that still an open issue somewhere, or has it been decided that aliasing `Box` is simply illegal and will remain that way?

**Comment by RalfJung** [comment]
I created https://github.com/rust-lang/unsafe-code-guidelines/issues/258 for the aliasing rules for Box.

**Comment by JakobDegen** [comment]
@RalfJung thoughts on leaving this open? We closed #76 . Maybe it's useful to have an issue about what the weaker validity invariants around box could be, but that would be premised on the answer to #258 being sufficiently weak

**Comment by RalfJung** [comment]
Some other issues that surely are relevant:
- https://github.com/rust-lang/unsafe-code-guidelines/issues/326
- https://github.com/rust-lang/unsafe-code-guidelines/issues/166
- https://github.com/rust-lang/unsafe-code-guidelines/issues/413

There are at least two questions which currently don't have another issue tracking them:
- Does `Box` have to be aligned? If the answer is "yes" then there already is magic involved since `NonNull` obviously doesn't have to be aligned.
- Is there any kind of derefencability requirement?

Basically there are a whole lot of axes and along each of them we can ask whether Box behaves like a reference or like `NonNull`, or maybe something in between (such as `Unique`). Right now for Miri the answer is basically always "like a mutable reference", except to account for "Box can be deallocated while the function is running". Many people consider that too strong. I am fairly sure that we do want alignment niches for Box though? So, I doubt the end result will be "like NonNull".
